src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1515 @@
+            bs = hidden_states.shape[0]
+            qkv = qkv.view(-1, 3, self.num_heads, self.head_dim)
+
+            attn_kwargs.update(
+                {
+                    "local_attention": self.local_attention,
+                    "cu_seqlens": cu_seqlens,
+                    "max_seqlen": max_seqlen,
+                    "bs": bs,
+                }
+            )
+        else:
+            bs, seqlen = hidden_states.shape[:2]
+            qkv = qkv.view(bs, seqlen, 3, self.num_heads, self.head_dim)
+
+            attn_kwargs.update(
+                {
+                    "position_ids": position_ids,
+                    "attention_mask": attention_mask,
+                    "bs": bs,
+                    "seqlen": seqlen,
+                }
+            )
+
+        if output_attentions:
+            if self.config._attn_implementation == "sdpa":
+                logger.warning_once(
+                    "Outputting attentions is only supported with the 'eager' attention implementation, "
+                    'not with "sdpa". Falling back to `attn_implementation="eager"`.'
+                )
+                self.config._attn_implementation = "eager"
+            elif self.config._attn_implementation != "eager":
+                logger.warning_once(
+                    "Outputting attentions is only supported with the eager attention implementation, "
+                    f'not with {self.config._attn_implementation}. Consider setting `attn_implementation="eager"`.'
+                    " Setting `output_attentions=False`."
+                )
+
+        attn_outputs = MODERNBERT_ATTENTION_FUNCTION[self.config._attn_implementation](
+            self,
+            qkv=qkv,
+            rotary_emb=self.rotary_emb,
+            **attn_kwargs,
+        )
+        hidden_states = attn_outputs[0]
+
+        return (self.out_drop(self.Wo(hidden_states)),) + attn_outputs[1:]  # add attentions if outputted
+
+
+class ModernBertEncoderLayer(nn.Module):
+    def __init__(self, config: ModernBertConfig, layer_id: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        if config.skip_first_prenorm and config.embedding_norm and layer_id == 0:
+            self.attn_norm = nn.Identity()
+        else:
+            self.attn_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.attn = ModernBertAttention(config=config, layer_id=layer_id)
+        self.mlp_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.mlp = ModernBertMLP(config)
+
+    def _init_weights(self, reset_params: bool = False):
+        if reset_params:
+            self.attn_norm.reset_parameters()
+            self.mlp_norm.reset_parameters()
+
+    @torch.compile(dynamic=True)
+    def compiled_mlp(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        return self.mlp(self.mlp_norm(hidden_states))
```

Comment:
okay, I like this but I want to make sure: 
1. this does have an upside in perf? 
2. this works on `mps`, and any device!

Reply:
I'd like for these `torch.compile` options to be optional (they were optional once too, we can reintroduce it), based on whether `torch.compile` can actually work (I believe it's not always possible on Windows for example).

Reply:
They need to be optional because torch.compile with dynamic shapes isn't compatible with DDP/FSDP/gradient checkpointing for all versions of PyTorch. Tom ported them from the research codebase because fusing the MLP norm into the MLP layer results in small numerical differences when in eager mode.

@ArthurZucker To answer your questions:
1. It's ~10% faster during pretraining
2. It's not [currently compatible with CPUs](https://github.com/triton-lang/triton-cpu), or MPS. AMD & Intel GPUs depend on the Triton version.

Reply:
Made it optional in 37a60302048150d12fbf4cdc3691317c3a40ce21. It's a parameter defaulting to `None`, and if it stays `None`, then it's set based on whether `triton` is available. This seems to work very well on Windows, allowing me to use both no `triton` for no compilation and `triton` (via https://github.com/woct0rdho/triton-windows) for compilation.

Reply:
In general @tomaarsen it's the kind of thing we avoid, as we don't have a great way to do this today. We can merge this for now as we are in a rush, but we will be needing a refactoring later!